File: js/player/PlayerHealth.js
// PlayerHealth.js - Health system with regen
class PlayerHealth {
    constructor() {
        this.max = 100;
        this.current = 100;
        this.regenRate = 2;
        this.regenDelay = 5;
        this.lastDamageTime = 0;
        this.isDead = false;
        this.isCritical = false;
        this.criticalThreshold = 25;
        this.heartbeatTimer = 0;
        this.onDeath = null;
        this.onDamage = null;
    }

    get percent() { return this.current / this.max; }

    takeDamage(amount) {
        if (this.isDead) return;
        this.current = Math.max(0, this.current - amount);
        this.lastDamageTime = performance.now() / 1000;

        audioManager.playSFX('player_hurt', 0.6);

        if (this.onDamage) this.onDamage(amount);

        if (this.current <= 0) {
            this.die();
        }
    }

    heal(amount) {
        if (this.isDead) return;
        this.current = Math.min(this.current + amount, this.max);
    }

    die() {
        this.isDead = true;
        this.current = 0;
        audioManager.playSFX('player_death', 0.8);
        if (this.onDeath) this.onDeath();
    }

    revive(healthPercent = 0.5) {
        this.isDead = false;
        this.current = this.max * healthPercent;
    }

    update(deltaTime) {
        if (this.isDead) return;

        const now = performance.now() / 1000;

        // Regen after delay
        if (now - this.lastDamageTime >= this.regenDelay && this.current < this.max) {
            this.current = Math.min(this.current + this.regenRate * deltaTime, this.max);
        }

        // Critical state
        this.isCritical = this.current <= this.criticalThreshold;
        if (this.isCritical) {
            this.heartbeatTimer += deltaTime;
            if (this.heartbeatTimer >= 1.0) {
                this.heartbeatTimer = 0;
                audioManager.playSFX('heartbeat', 0.5);
            }
        }
    }

    reset() {
        this.current = this.max;
        this.isDead = false;
        this.isCritical = false;
        this.heartbeatTimer = 0;
    }
}

File: Assets/Scripts/Player/PlayerHealth.cs
using UnityEngine;
using Mirror;
using System;

namespace ZombieSurvival.Player
{
    public class PlayerHealth : NetworkBehaviour
    {
        [Header("Health Settings")]
        [SerializeField] private float maxHealth = 100f;
        [SerializeField] private float healthRegenRate = 1f;
        [SerializeField] private float regenDelay = 5f;
        [SerializeField] private float criticalHealthThreshold = 25f;

        [Header("Damage Effects")]
        [SerializeField] private float damageScreenDuration = 0.5f;
        [SerializeField] private float heartbeatInterval = 1f;

        [SyncVar(hook = nameof(OnHealthChanged))]
        private float currentHealth;

        [SyncVar] private bool isDead;

        private float lastDamageTime;
        private float heartbeatTimer;

        public float CurrentHealth => currentHealth;
        public float MaxHealth => maxHealth;
        public float HealthPercent => currentHealth / maxHealth;
        public bool IsDead => isDead;
        public bool IsCritical => currentHealth <= criticalHealthThreshold;

        public event Action<float, float> OnHealthUpdated; // current, max
        public event Action OnDeath;
        public event Action OnRevived;
        public event Action<float> OnDamageTaken;
        public event Action OnCriticalHealth;

        public override void OnStartServer()
        {
            base.OnStartServer();
            currentHealth = maxHealth;
            isDead = false;
        }

        public override void OnStartClient()
        {
            base.OnStartClient();
            OnHealthUpdated?.Invoke(currentHealth, maxHealth);
        }

        private void Update()
        {
            if (!isServer) return;

            if (!isDead && currentHealth < maxHealth)
            {
                if (Time.time - lastDamageTime >= regenDelay)
                {
                    currentHealth = Mathf.Min(currentHealth + healthRegenRate * Time.deltaTime, maxHealth);
                }
            }

            if (isLocalPlayer && IsCritical && !isDead)
            {
                heartbeatTimer += Time.deltaTime;
                if (heartbeatTimer >= heartbeatInterval)
                {
                    heartbeatTimer = 0;
                    AudioManager.AudioManager.Instance?.PlaySFX("heartbeat");
                }
            }
        }

        [Server]
        public void TakeDamage(float damage, GameObject attacker = null)
        {
            if (isDead) return;

            currentHealth = Mathf.Max(0, currentHealth - damage);
            lastDamageTime = Time.time;

            RpcOnDamageTaken(damage);

            if (currentHealth <= 0)
            {
                Die(attacker);
            }
            else if (IsCritical)
            {
                RpcOnCriticalHealth();
            }
        }

        [Server]
        public void Heal(float amount)
        {
            if (isDead) return;
            currentHealth = Mathf.Min(currentHealth + amount, maxHealth);
        }

        [Server]
        private void Die(GameObject killer)
        {
            isDead = true;
            currentHealth = 0;

            if (killer != null)
            {
                var killerRanking = killer.GetComponent<Ranking.PlayerScore>();
                if (killerRanking != null)
                {
                    killerRanking.AddKill();
                }
            }

            var myScore = GetComponent<Ranking.PlayerScore>();
            if (myScore != null)
            {
                myScore.AddDeath();
            }

            RpcOnDeath();
        }

        [Server]
        public void Revive(float healthPercent = 0.5f)
        {
            isDead = false;
            currentHealth = maxHealth * healthPercent;
            RpcOnRevived();
        }

        [ClientRpc]
        private void RpcOnDamageTaken(float damage)
        {
            OnDamageTaken?.Invoke(damage);
            AudioManager.AudioManager.Instance?.PlaySFX("player_hurt");
        }

        [ClientRpc]
        private void RpcOnDeath()
        {
            OnDeath?.Invoke();
            AudioManager.AudioManager.Instance?.PlaySFX("player_death");
        }

        [ClientRpc]
        private void RpcOnRevived()
        {
            OnRevived?.Invoke();
        }

        [ClientRpc]
        private void RpcOnCriticalHealth()
        {
            OnCriticalHealth?.Invoke();
        }

        private void OnHealthChanged(float oldHealth, float newHealth)
        {
            OnHealthUpdated?.Invoke(newHealth, maxHealth);
        }
    }
}
